File: laravel/app/Http/Controllers/Api/Content/ArticleCommentController.php
<?php

namespace App\Http\Controllers\Api\Content;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use Illuminate\Http\JsonResponse;
use App\Models\Content\ArticleComment;
use App\Models\Content\Article;
use App\Models\Players\Player;

class ArticleCommentController extends Controller
{
    public function index(): JsonResponse
    {
        return response()->json(ArticleComment::all());
    }

    public function store(Request $request): JsonResponse
    {
        $validated = $request->validate([
            'body' => 'required|string|max:200',
            'is_hidden' => 'required|boolean|max:200',
            'created_at' => 'required|date|max:200',
            'article_id' => 'nullable|exists:articles,id',
            'author_id' => 'required|exists:players,id',
            'parent_comment_id' => 'nullable|exists:article_comments,id',
        ]);
        $item = ArticleComment::create($validated);
        return response()->json($item, 201);
    }

    public function show(ArticleComment $articleComment): JsonResponse
    {
        return response()->json($articleComment);
    }

    public function update(Request $request, ArticleComment $articleComment): JsonResponse
    {
        $validated = $request->validate([
            'body' => 'sometimes|nullable|string|max:200',
            'is_hidden' => 'sometimes|nullable|boolean|max:200',
            'created_at' => 'sometimes|nullable|date|max:200',
            'article_id' => 'sometimes|nullable|exists:articles,id',
            'author_id' => 'sometimes|nullable|exists:players,id',
            'parent_comment_id' => 'sometimes|nullable|exists:article_comments,id',
        ]);
        $articleComment->update($validated);
        return response()->json($articleComment);
    }

    public function destroy(ArticleComment $articleComment): JsonResponse
    {
        $articleComment->delete();
        return response()->json(null, 204);
    }
}

File: phoenix/lib/cards_project_web/controllers/content/article_comment_controller.ex
defmodule CardsProjectWeb.Content.ArticleCommentController do
  use Phoenix.Controller, formats: [:json]
  import Plug.Conn

  alias CardsProject.Content
  alias CardsProject.Content.ArticleComment

  def index(conn, _params) do
    article_comments = Content.list_article_comments()
    json(conn, Enum.map(article_comments, &serialize_article_comment/1))
  end

  def show(conn, %{"id" => id}) do
    article_comment = Content.get_article_comment!(id)
    json(conn, serialize_article_comment(article_comment))
  end

  def create(conn, params) do
    case Content.create_article_comment(params) do
      {:ok, article_comment} ->
        conn
        |> put_status(:created)
        |> json(serialize_article_comment(article_comment))

      {:error, changeset} ->
        conn
        |> put_status(:unprocessable_entity)
        |> json(%{errors: format_errors(changeset)})
    end
  end

  def update(conn, %{"id" => id} = params) do
    article_comment = Content.get_article_comment!(id)
    case Content.update_article_comment(article_comment, params) do
      {:ok, article_comment} ->
        json(conn, serialize_article_comment(article_comment))

      {:error, changeset} ->
        conn
        |> put_status(:unprocessable_entity)
        |> json(%{errors: format_errors(changeset)})
    end
  end

  def delete(conn, %{"id" => id}) do
    article_comment = Content.get_article_comment!(id)
    Content.delete_article_comment(article_comment)
    send_resp(conn, :no_content, "")
  end

  # POST /api/comments/{id}/hide
  def hide(conn, %{"id" => id}) do
    Content.article_comment_hide_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # POST /api/comments/{id}/unhide
  def unhide(conn, %{"id" => id}) do
    Content.article_comment_unhide_behavior(id)
    send_resp(conn, :no_content, "")
  end

  defp serialize_article_comment(%ArticleComment{} = record) do
    Map.take(record, [:id, :body, :is_hidden, :created_at, :article_id, :author_id, :parent_comment_id])
  end

  defp format_errors(changeset) do
    Ecto.Changeset.traverse_errors(changeset, fn {msg, opts} ->
      Regex.replace(~r"%{(\w+)}", msg, fn _, key ->
        opts |> Keyword.get(String.to_existing_atom(key), key) |> to_string()
      end)
    end)
  end
end
